src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,117 @@
 package com.es.phoneshop.model.product;
 
-import java.util.List;
+import java.util.*;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private static ArrayListProductDao instance;
+    private Long maxId;
+    private List<Product> products;
+    private final Object lock = new Object();
+
+    private ArrayListProductDao() {
+        products = new ArrayList<>();
+        maxId = 0L;
+    }
+
+    public static synchronized ArrayListProductDao getInstance() {
+        if (instance == null) {
+            instance = new ArrayListProductDao();
+        }
+        return instance;
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public Product getProduct(Long id) throws ProductNotFoundException {
+        if (id == null) {
+            throw new ProductNotFoundException(id);
+        }
+        synchronized (lock) {
+            return products.stream()
+                    .filter(product -> id.equals(product.getId()))
+                    .findAny()
+                    .orElseThrow(() -> new ProductNotFoundException(id));
+        }
+    }
+
+    private int compare(Product p1, Product p2, SortField sortField, SortOrder sortOrder) {
+        int compareDiff = 0;
+        if (sortField == SortField.price) {
+            compareDiff = p1.getPrice().compareTo(p2.getPrice());
+        } else if (sortField == SortField.description) {
+            compareDiff = p1.getDescription().compareTo(p2.getDescription());
+        }
+        return SortOrder.desc == sortOrder ? -compareDiff : compareDiff;
+    }
+
+    private List<Product> findByQuery(String query) {
+        String[] words = query.split("\\s");
+        synchronized (lock) {
+            return products.stream()
+                    .collect(Collectors.toMap(product -> product,
+                            product -> Arrays.stream(words)
+                                    .filter(word -> product
+                                            .getDescription()
+                                            .toLowerCase()
+                                            .contains(word.toLowerCase()))
+                                    .count()))
```

Comment:
Looks quite monstrous. Try to move some logic to separate method.
Btw, there is `Function.identity()` that can be used when you need to do something like `product -> product`